client/src/templates/Challenges/utils/frame.ts

```diff
@@ -106,60 +111,72 @@ export const runTestInTestFrame = async function (
       frame.title = title;
     }
     return {
-      ...ctx,
+      ...frameContext,
       element: frame
     };
   };
 
 const hiddenFrameClassName = 'hide-test-frame';
-const mountFrame = (document: Document, id: string) => (ctx: Context) => {
-  const { element }: { element: HTMLIFrameElement } = ctx;
-  const oldFrame = document.getElementById(element.id) as HTMLIFrameElement;
-  if (oldFrame) {
-    element.className = oldFrame.className || hiddenFrameClassName;
-    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
-    oldFrame.parentNode!.replaceChild(element, oldFrame);
-    // only test frames can be added (and hidden) here, other frames must be
-    // added by react
-  } else if (id === testId) {
-    element.className = hiddenFrameClassName;
-    document.body.appendChild(element);
-  }
-  return {
-    ...ctx,
-    element,
-    document: element.contentDocument,
-    window: element.contentWindow
+const mountFrame =
+  (document: Document, id: string) => (frameContext: Context) => {
+    const { element }: { element: HTMLIFrameElement } = frameContext;
+    const oldFrame = document.getElementById(element.id) as HTMLIFrameElement;
+    if (oldFrame) {
+      element.className = oldFrame.className || hiddenFrameClassName;
+      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+      oldFrame.parentNode!.replaceChild(element, oldFrame);
+      // only test frames can be added (and hidden) here, other frames must be
+      // added by react
+    } else if (id === testId) {
+      element.className = hiddenFrameClassName;
+      document.body.appendChild(element);
+    }
+    return {
+      ...frameContext,
+      element,
+      document: element.contentDocument,
+      window: element.contentWindow
+    };
   };
-};
 
-const buildProxyConsole = (proxyLogger?: ProxyLogger) => (ctx: Context) => {
-  // window does not exist if the preview is hidden, so we have to check.
-  if (proxyLogger && ctx?.window) {
-    const oldLog = ctx.window.console.log.bind(ctx.window.console);
-    ctx.window.console.log = function proxyConsole(...args: string[]) {
-      proxyLogger(args.map((arg: string) => utilsFormat(arg)).join(' '));
-      return oldLog(...(args as []));
-    };
-  }
-  return ctx;
-};
+const buildProxyConsole =
+  (proxyLogger?: ProxyLogger) => (frameContext: Context) => {
+    // window does not exist if the preview is hidden, so we have to check.
+    if (proxyLogger && frameContext?.window) {
+      const oldLog = frameContext.window.console.log.bind(
+        frameContext.window.console
+      );
+      frameContext.window.console.log = function proxyConsole(
+        ...args: string[]
+      ) {
+        proxyLogger(args.map((arg: string) => utilsFormat(arg)).join(' '));
+        return oldLog(...(args as []));
+      };
 
-const initTestFrame = (frameReady?: () => void) => (ctx: Context) => {
-  waitForFrame(ctx)
+      frameContext.window.i18next = i18next;
```

Comment:
This seems a little out of place, since it's nothing to do with the proxyLogger.  How about putting it in it's own function?  Also, we can reduce the chance of weird collisions by calling, say `__i18next`.

Reply:
I couldn't change it to `__i18next` typescript wasn't happy with it, so I change it to `i18nContent`.

About moving it to its own function, did you mean something like this?

```ts
const buildProxyConsole =
  (proxyLogger?: ProxyLogger) => (frameContext: Context) => {
    // window does not exist if the preview is hidden, so we have to check.
    if (proxyLogger && frameContext?.window) {
      const oldLog = frameContext.window.console.log.bind(
        frameContext.window.console
      );
      frameContext.window.console.log = function proxyConsole(
        ...args: string[]
      ) {
        proxyLogger(args.map((arg: string) => utilsFormat(arg)).join(' '));
        return oldLog(...(args as []));
      };
    }

    frameContext.window.i18nContent = i18next;

    return frameContext;
  };
```

I will be honest, I don't see another option, so I may need more details 😓

Reply:
That's just eslint being overly pick.  Feel free to ignore the naming conventions if you want. I think that rule is probably more trouble than it's worth.

> About moving it to its own function, did you mean something like this?

Not exactly. The thing is this function `buildProxyConsole` doesn't have the best name, but it at least does attach the proxy console to the window object (replacing console.log).  It doesn't have anything to do with `i18n`, though, so it would be surprising to discover that it did something i18n related.

To be honest, since this already has a bad name, why not call this whole function `updateWindowGlobals` instead of creating yet another function? I think that's clear enough and means we don't need yet another function.

Finally, I'd return early if frameContent.window doesn't exist, since you cannot do anything without it.

Reply:
I did like your second suggestion, and I want to implement it. but with multiple functions changing `frameContent`, I couldn't didn't take it, because it would have been weird if there is a function that called that while there are multiple functions doing the same thing.